### PR TITLE
feat(project): add stylelint-plugin-carbon-tokens package

### PR DESCRIPTION
**Pull request by lee-chase**
Closesmight be putting it a bit strong but relates to #5823

Donate the package https://github.com/lee-chase/carbon-use to the mono repo.

The linter looks at type, theme (including color), layout and motion and tries to determine whether or not carbon tokens are being used. It relies on javascript and SCSS files in other carbon packages to achieve this.

NOTE 1: There may well be SCSS value syntax this stylelint plugin fails to recognise. When this happens it should emit a console.warn message requesting the user raises an issue. It does not try to raise a linter warning as some teams treat warnings as errors.
NOTE 2: It would be preferable to replace parsing of SCSS files to identify tokens with direct import from tokens.js files. Where tokens are exported as "unstable__tokens" SCSS parsing is currently prefered.

#### Changelog

**New**

- Add `stylelint-carbon-use` package

**Changed**

Setup was changed to add the global function found in packages/stylelint-carbon-use as there is no way I know of to use local config in a mono repo. https://github.com/facebook/jest/issues/5921

M	tasks/jest/setup.js
M	yarn.lock

#### Testing / Reviewing

Testing so far has been through the use of stylelint-carbon-use in an internal project.  Can provide link to IBMers only.

### CODE REVIEW

**Comment by lee-chase** [comment]
NOTE: I expect to continue to develop and maintain this package as a side project and will be using it in my daily work.

**Comment by lee-chase** [comment]
A bit stuck on the above CircleCI failure I have yarn 1.22.4 and a fresh copy of the repo.

**Comment by joshblack** [comment]
@lee-chase how would you feel about the name being something like `stylelint-plugin-carbon` or `stylelint-plugin-carbon-tokens`?

**Comment by lee-chase** [comment]
> @lee-chase how would you feel about the name being something like `stylelint-plugin-carbon` or `stylelint-plugin-carbon-tokens`?

Happy with either, torn between the two. 

The linter allows for tokens and some functions, which makes me lean towards the first. However, the term people have picked up internally in IBM is "tokens".

Do you have a preference @joshblack?

**Comment by lee-chase** [comment]
Actually @joshblack thinking about it I prefer `stylelint-plugin-carbon-tokens` for the recognition it will get from people asked to use tokens and because any functions are simply alternative values for tokens.

Once merged it would be good to discuss versioning and publishing. If it were my choice I would probably go with 1.0.0-alpha.0 as the main thing it needs is more users and people may see 0.1.0 as too early to use.

Happy to take guidance on the typical path of Carbon packages.

File: packages/stylelint-plugin-carbon-tokens/README.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+> A stylelint plugin to support the use of carbon component tokens.
+
+## Getting started
+
+To install `stylelint-plugin-carbon-tokens` in your project, you will need to
+run the following command using [npm](https://www.npmjs.com/):
+
+```bash
+npm install -S stylelint-plugin-carbon-tokens
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/README.md]: this flag is redundant now isn't it? or should this be a dev dependency?

File: packages/stylelint-plugin-carbon-tokens/README.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+```bash
+yarn add stylelint-plugin-carbon-tokens
+```
+
+## Usage
+
+This project is intended to help users identify cases where tokens, functions
+and mixins defined as part of various projects relating to the Carbon Design
+System are used. [www.carbondesignsystem.com]
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/README.md]: is this supposed to be a raw link or should it have a title?

File: packages/stylelint-plugin-carbon-tokens/README.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+It consists of a stylelint plugin and depends on various Carbon Design System
+packages for loading settings.
+
+It includes, but may not be limited to, linting for @carbon/themes,
+@carbon/colors, @carbon/layout, @carbon/type and @carbon/motion.
+
+Not incluced, as they're not used through SCSS, are Carbon Icons, Grid and any
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/README.md]:
```suggestion
Not included, as they're not used through SCSS, are Carbon Icons, Grid and any
```

File: packages/stylelint-plugin-carbon-tokens/README.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+other DOM related checks..
+
+### Please be helpful
+
+Before we start this project is a work in progress which deliberately outputs
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/README.md]:
```suggestion
Before we start, this project is a work in progress which deliberately outputs
```

File: packages/stylelint-plugin-carbon-tokens/README.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+warnings when it comes accross a syntax that has not yet been catered for. If
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/README.md]:
```suggestion
warnings when it comes across syntax that has not yet been catered for. If
```

File: packages/stylelint-plugin-carbon-tokens/README.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+The last option here shows how you could elect to check your own tokens refer to
+values acceptable to the linter.
+
+- `acceptValues: ["$/^\\$my-color--/"]` - Accept SCSS variables starting
+  "\\\$my-color--"
+
+#### includeProps Range
+
+Can innclude a range value expressed inside greater than and less than signs.
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/README.md]:
```suggestion
Can include a range value expressed inside greater than and less than signs.
```

File: packages/stylelint-plugin-carbon-tokens/README.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+- If no range is specified the whole value list is checked.
+
+- A single value means only that value in a list is checked
+
+- Two values represent start and end values of a range in the list.
+
+#### includeProps specific values
+
+For some props e.g. transform we are only intereste in values that match a
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/README.md]:
```suggestion
For some props e.g. transform we are only interested in values that match a
```

File: packages/stylelint-plugin-carbon-tokens/README.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+certain criteria.
+
+This is specified as part of the includedProp inside \[].
+
+e.g. `translate[/^transform/]`
+
+In this case only values starting `transform` are tested so not `skew` for
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/README.md:203]:
```suggestion
In this case only values starting with `transform` are tested, so not `skew` for
```

File: packages/stylelint-plugin-carbon-tokens/src/rules/layout-token-use/README.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
+# Layout Token Use
+
+This rule is intended enfoce use of Carbon theme tokens, functions, mixins and
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/src/rules/layout-token-use/README.md]:
```suggestion
This rule is intended to enforce the use of Carbon theme tokens, functions, mixins and
```

File: packages/stylelint-plugin-carbon-tokens/src/rules/layout-token-use/README.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    '/^padding-/',
+    'height',
+    'width',
+    'left',
+    'top',
+    'bottom',
+    'right',
+    'transform',
+  ],
+  // Accept transparent, common reset values, 0, proportioanl values,
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/src/rules/layout-token-use/README.md]:
```suggestion
  // Accept transparent, common reset values, 0, proportional values,
```

File: packages/stylelint-plugin-carbon-tokens/src/rules/motion-token-use/README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Motion Token Use
+
+This rule is intended enfoce use of Carbon theme tokens, functions, mixins and
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/src/rules/motion-token-use/README.md]:
```suggestion
This rule is intended to enforce the use of Carbon theme tokens, functions, mixins and
```

File: packages/stylelint-plugin-carbon-tokens/src/rules/theme-token-use/README.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
+# Theme Token Use
+
+This rule is intended enfoce use of Carbon theme tokens, functions, mixins and
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/src/rules/theme-token-use/README.md:3]:
```suggestion
This rule is intended to enforce the use of Carbon theme tokens, functions, mixins and
```

File: packages/stylelint-plugin-carbon-tokens/src/rules/type-token-use/README.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# Type Token Use
+
+This rule is intended enfoce use of Carbon type tokens, functions, mixins and
```
Comment (emyarod) [inline packages/stylelint-plugin-carbon-tokens/src/rules/type-token-use/README.md]:
```suggestion
This rule is intended to enforce the use of Carbon type tokens, functions, mixins and
```

File: packages/stylelint-plugin-carbon-tokens/README.md (hunk 1 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/README.md]: Replaced with -D

File: packages/stylelint-plugin-carbon-tokens/README.md (hunk 2 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/README.md]: Changed to
```
see [www.carbondesignsystem.com](www.carbondesignsystem.com) 
```

Oddly we are not very self-referential in source, can change to something more appropriate if desired.

File: packages/stylelint-plugin-carbon-tokens/README.md (hunk 3 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/README.md]: Corrected spelling and installed spell checker extension. Do you have a favourite spelling extension for VSCode?

File: packages/stylelint-plugin-carbon-tokens/README.md (hunk 4 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/README.md]: Corrected grammar.

File: packages/stylelint-plugin-carbon-tokens/README.md (hunk 5 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/README.md]: Corrected spelling and grammar, wonder if Grammarly has a VSCode plugin ;-)

File: packages/stylelint-plugin-carbon-tokens/src/rules/layout-token-use/README.md (hunk 9 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/src/rules/layout-token-use/README.md]: Agreed and updated.

File: packages/stylelint-plugin-carbon-tokens/src/rules/layout-token-use/README.md (hunk 10 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/src/rules/layout-token-use/README.md]: Agreed and updated.

File: packages/stylelint-plugin-carbon-tokens/src/rules/theme-token-use/README.md (hunk 12 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/src/rules/theme-token-use/README.md:3]: Agreed and updated.

File: packages/stylelint-plugin-carbon-tokens/src/rules/type-token-use/README.md (hunk 13 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/src/rules/type-token-use/README.md]: Agreed and updated.

File: packages/stylelint-plugin-carbon-tokens/src/rules/motion-token-use/README.md (hunk 11 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/src/rules/motion-token-use/README.md]: Agreed and updated.

File: packages/stylelint-plugin-carbon-tokens/README.md (hunk 6 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/README.md]: Corrected

File: packages/stylelint-plugin-carbon-tokens/README.md (hunk 7 above)
Comment (lee-chase) [reply packages/stylelint-plugin-carbon-tokens/README.md]: Corrected

**Comment by lee-chase** [comment]
Apologies @emyarod my brain only processed the [Commit suggestion] button after I had pushed updates.

File: packages/stylelint-plugin-carbon-tokens/README.md (hunk 3 above)
Comment (tw15egan) [reply packages/stylelint-plugin-carbon-tokens/README.md]: I've been using https://github.com/streetsidesoftware/vscode-spell-checker

**Comment by joshblack** [comment]
One quick question for you @lee-chase what would the recommended options be for the plugin? We might be able to have a `recommended.js` file with them enabled and could try it out in the monorepo 👀

**Comment by lee-chase** [comment]
> One quick question for you @lee-chase what would the recommended options be for the plugin? We might be able to have a `recommended.js` file with them enabled and could try it out in the monorepo 👀

@joshblack currently using this in internal repository 

```module.exports = {
  root: true,
  extends: ['stylelint-config-carbon'],
  plugins: ['stylelint-carbon-use'],
  rules: {
    // ADDED TO TEST CARBON USE
    'carbon/layout-token-use': [true, { severity: 'warning' }],
    'carbon/motion-token-use': [true, { severity: 'warning' }],
    'carbon/theme-token-use': [true, { severity: 'warning' }],
    'carbon/type-token-use': [
      true,
      {
        severity: 'warning',
        acceptCarbonTypeScaleFunction: true,
        acceptCarbonFontWeightFunction: true,
      },
    ],
  },
};
```

We could choose more of the 'acceptSomeSimpleOption' settings as recommended, and there may be some you think are missing. I would not envisage recommends.js including any includeProps or acceptProps settings.

The full list of current acceptSomeSimpleOptions and their default values is:

```
// layout
  acceptUndefinedVariables: true,
  acceptContainerTokens: false,
  acceptIconSizeTokens: false,
  acceptFluidSpacingTokens: false,
  acceptCarbonMiniUnitsFunction: false,
// motion
  acceptUndefinedVariables: true,
// theme
  acceptCarbonColorTokens: false,
  acceptIBMColorTokens: false,
  acceptUndefinedVariables: true,
// type
  acceptCarbonFontWeightFunction: false, 
  acceptCarbonTypeScaleFunction: false, 
```

The most controversial of these may well be acceptUndefinedVariables. Essentially the plan is to assume they're valid and add an include props pattern. Hmm, maybe it should default to false and be called out in the README separately from the 'Advanced usage' section.

**Comment by joshblack** [comment]
Unfortunately, it seems like there isn't much activity from our end for this PR, sorry about that @lee-chase 😞 

We really appreciate you taking the time to make this PR, instead of being in the monorepo it might make sense to have this live either in the cloud and cognitive project or as a standalone repo as a result.

Thanks again for taking the time hear, and sorry again about the delay. I'm going to go ahead and close this out but feel free to comment or message me and anyone on the team would be happy to re-open 👍

**Pull request closed without merging**